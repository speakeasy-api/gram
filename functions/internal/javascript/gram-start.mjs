import { existsSync } from "node:fs";
import { open } from "node:fs/promises";
import process from "node:process";
import path from "node:path";
import url from "node:url";
import { Writable } from "node:stream";

export const ERROR_CODES = /** @type {const} */ ({
  UNEXPECTED: "gram_err_000",
  INVALID_TOOL_RESULT: "gram_err_001",
  TOOL_CALL_FAILED: "gram_err_002",
  IMPORT_FAILURE: "gram_err_003",
  INVALID_TOOL_FUNC: "gram_err_004",
});

class FunctionsError extends Error {
  /**
   * @param {typeof ERROR_CODES[keyof typeof ERROR_CODES]} code
   * @param {string} message
   * @param {string | undefined} [cause]
   */
  constructor(code, message, cause) {
    super(`${message} (${code})`);
    this.name = "FunctionsError";
    this.cause = cause;
    this.code = code;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      cause: this.cause,
    };
  }
}

/**
 * @param {string[]} args
 * @returns {{pipePath: string, toolCall: {name: string, input: unknown}}}
 */
function parseArgs(args) {
  args = args.slice(2);

  if (args.length !== 2) {
    throw new Error(
      "Expected two command-line argument but got " + args.length
    );
  }

  const pipePath = args[0];
  if (typeof pipePath !== "string" || !existsSync(pipePath)) {
    throw new Error(`Named pipe does not exist: ${pipePath}`);
  }

  const callArg = args[1];
  if (typeof callArg !== "string") {
    throw new Error(
      `Invalid tool call argument type: expected string, got ${typeof callArg}`
    );
  }

  const toolCall = JSON.parse(callArg);

  if (typeof toolCall !== "object" || toolCall === null) {
    throw new Error("Tool call argument must be a valid JSON object");
  }

  if (typeof toolCall.name !== "string") {
    throw new Error("Argument must have a string 'name' property");
  }

  return { pipePath, toolCall };
}

/**
 * @param {(name: string, input: unknown) => Promise<Response>} func
 * @param {string} name
 * @param {unknown} input
 * @returns {Promise<Response>}
 */
async function callTool(func, name, input) {
  try {
    const response = await func(name, input);
    if (!(response instanceof Response)) {
      throw new FunctionsError(
        ERROR_CODES.INVALID_TOOL_RESULT,
        "Tool call did not return a valid response",
        `Expected instance of \`Response\` but got \`${typeof response}\``
      );
    }
    return response;
  } catch (e) {
    if (e instanceof FunctionsError) {
      throw e;
    } else {
      let msg = e instanceof Error ? e.message : String(e);
      msg = msg || "Tool call failed";
      throw new FunctionsError(ERROR_CODES.TOOL_CALL_FAILED, msg);
    }
  }
}

/**
 *
 * @param {import("node:fs/promises").FileHandle} pipeFile
 * @param {FunctionsError} error
 */
async function writeFunctionsError(pipeFile, error) {
  const text = JSON.stringify(error.toJSON());

  return writeHTTPResponse(
    pipeFile,
    new Response(text, {
      status: 500,
      headers: {
        "Content-Type": "application/json",
        "Content-Length": String(Buffer.byteLength(text, "utf8")),
        "Gram-Functions-Error": error.code,
      },
    })
  );
}

/**
 * @param {import("node:fs/promises").FileHandle} pipeFile
 * @param {Response} response
 */
async function writeHTTPResponse(pipeFile, response) {
  const status = response.status || 200;
  const statusText = response.statusText || "";

  await pipeFile.write(`HTTP/1.1 ${status} ${statusText}\r\n`);

  for (const [key, value] of response.headers) {
    await pipeFile.write(`${key}: ${value}\r\n`);
  }

  await pipeFile.write("\r\n");

  if (!response.body) {
    return;
  }

  const dest = pipeFile.createWriteStream({
    autoClose: true,
    emitClose: true,
    flush: true,
  });

  await response.body.pipeTo(Writable.toWeb(dest));
}

/**
 * @param {string} codePath
 * @returns {Promise<{ok: true, value: (name: string, input: unknown) => Promise<Response>} | {ok: false, error: FunctionsError}>}
 */
async function importToolCallHandler(codePath) {
  try {
    const { handleToolCall: f } = await import(codePath).catch((e) => {
      const filename = path.basename(codePath);
      throw new FunctionsError(
        ERROR_CODES.IMPORT_FAILURE,
        "Unable to import user code",
        `Failed to import ${filename}: ${
          e instanceof Error ? e.message : String(e)
        }`
      );
    });

    if (typeof f !== "function") {
      const filename = path.basename(codePath);

      throw new FunctionsError(
        ERROR_CODES.INVALID_TOOL_FUNC,
        "Unable to call tool",
        "handleToolCall function not found in " + filename
      );
    }

    return { ok: true, value: f };
  } catch (e) {
    if (e instanceof FunctionsError) {
      return { ok: false, error: e };
    } else {
      return {
        ok: false,
        error: new FunctionsError(
          ERROR_CODES.UNEXPECTED,
          "Unexpected error occurred",
          e instanceof Error ? e.message : String(e)
        ),
      };
    }
  }
}

const USER_CODE_PATH = path.join(process.cwd(), "functions.js");

/**
 *
 * @param {import("node:fs/promises").FileHandle} pipeFile
 * @param {{name: string, input: unknown}} toolCall
 * @param {string} codePath
 * @returns
 */
async function handleToolCall(pipeFile, toolCall, codePath) {
  const importResult = await importToolCallHandler(codePath);
  if (!importResult.ok) {
    await writeFunctionsError(pipeFile, importResult.error);
    return;
  }

  try {
    const res = await callTool(
      importResult.value,
      toolCall.name,
      toolCall.input
    );
    await writeHTTPResponse(pipeFile, res);
  } catch (e) {
    if (e instanceof FunctionsError) {
      return await writeFunctionsError(pipeFile, e);
    } else {
      throw e;
    }
  }
}

export async function main(args = process.argv, codePath = USER_CODE_PATH) {
  const { pipePath, toolCall } = parseArgs(args);

  const pipeFile = await open(pipePath, "w");
  try {
    await handleToolCall(pipeFile, toolCall, codePath);
  } finally {
    await pipeFile.close();
  }
}

if (import.meta.url.startsWith("file:")) {
  const modulePath = url.fileURLToPath(import.meta.url);
  if (process.argv[1] === modulePath) {
    main();
  }
}
