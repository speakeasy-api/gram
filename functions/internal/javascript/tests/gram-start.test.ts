import { mkdtemp, open, readFile } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { test, expect, beforeAll, afterEach, afterAll } from "vitest";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { ERROR_CODES, main } from "../gram-start.mjs";

async function fakepipe() {
  const dir = await mkdtemp(join(tmpdir(), "gramfunc-"));
  const responsePath = join(dir, "response.txt");
  const handle = await open(responsePath, "w");
  await handle.close();
  return responsePath;
}

const server = setupServer(
  ...[
    http.get("http://localhost:7357/pet", () => {
      return HttpResponse.json({ name: "Finn", type: "cat" });
    }),
    http.get("http://localhost:7357/logo", async () => {
      const logo = await readFile(join(import.meta.dirname, "logo.png"));
      return HttpResponse.arrayBuffer(logo.buffer, {
        headers: { "Content-Type": "image/png" },
      });
    }),
  ]
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("valid tool call with args", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "get-weather",
    input: { city: "San Francisco" },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();
});

test("valid tool call without args", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "ping",
    input: {},
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();
});

test("valid tool call adds custom headers", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "list-products",
    input: { cursor: "1" },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();

  expect(content).toMatch(/^x-next-cursor: 4$/im);
});

test("valid tool call explicit 4XX", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "create-charge",
    input: { charge: -100 },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();
});

test("valid tool call explicit 4XX", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "create-charge",
    input: { charge: -100 },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();
});

test("proxy good downstream fetch calls", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "proxy",
    input: { url: "http://localhost:7357/pet" },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toMatchSnapshot();

  const pet = JSON.parse(content.trim().split("\n").at(-1) ?? "");
  expect(pet).toEqual({ name: "Finn", type: "cat" });
});

test("proxy good downstream that returns binary data", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "proxy",
    input: { url: "http://localhost:7357/logo" },
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const text = await readFile(pipePath, "utf-8");

  // Get the content length of the body and then we'll be able to load the
  // result file again as a byte array and read the image data.
  const lhdr = text
    .trim()
    .split("\n")
    .find((line) => line.toLowerCase().startsWith("content-length:"));
  const lenstr = lhdr?.split(":")?.[1]?.trim() ?? "0";
  const len = parseInt(lenstr, 10);

  const binary = await readFile(pipePath);
  const imageData = binary.subarray(-len);
  const logo = await readFile(join(import.meta.dirname, "logo.png"));
  expect(imageData).toEqual(logo);
});

test("catches tool calls that throw", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "fail-tool",
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toContain(ERROR_CODES.TOOL_CALL_FAILED);
  expect(content).toMatchSnapshot();

  const res = JSON.parse(content.trim().split("\n").at(-1) ?? "");
  expect(res).toEqual({
    name: "FunctionsError",
    message: expect.stringMatching(/Intentional failure/),
  });
});

test("fails when tool call does not return Response", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "null-tool",
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "good.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toContain(ERROR_CODES.INVALID_TOOL_RESULT);
  expect(content).toMatchSnapshot();
});

test("fails when functions file does not exist", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "null-tool",
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "nonexistent.js")
  );

  const content = await readFile(pipePath, "utf-8");
  const res = JSON.parse(content.trim().split("\n").at(-1) ?? "");
  expect(res).toEqual({
    name: "FunctionsError",
    message: expect.stringMatching(
      /Unable to import user code \(gram_err_003\)/
    ),
    cause: expect.stringMatching(/^Failed to import nonexistent\.js/),
  });
});

test("fails when functions file does not export handleToolCall", async () => {
  const pipePath = await fakepipe();
  const args = JSON.stringify({
    name: "null-tool",
  });

  await main(
    ["node", "./gram-start.mjs", pipePath, args],
    join(import.meta.dirname, "empty.js")
  );

  const content = await readFile(pipePath, "utf-8");
  expect(content).toContain(ERROR_CODES.INVALID_TOOL_FUNC);
  expect(content).toMatchSnapshot();
});
