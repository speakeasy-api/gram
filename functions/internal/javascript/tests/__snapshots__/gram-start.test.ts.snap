// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`catches tool calls that throw 1`] = `
"HTTP/1.1 500 
content-length: 72
content-type: application/json
gram-functions-error: gram_err_002

{"name":"FunctionsError","message":"Intentional failure (gram_err_002)"}"
`;

exports[`fails when functions file does not export handleToolCall 1`] = `
"HTTP/1.1 500 
content-length: 128
content-type: application/json
gram-functions-error: gram_err_004

{"name":"FunctionsError","message":"Unable to call tool (gram_err_004)","cause":"handleToolCall function not found in empty.js"}"
`;

exports[`fails when tool call does not return Response 1`] = `
"HTTP/1.1 500 
content-length: 153
content-type: application/json
gram-functions-error: gram_err_001

{"name":"FunctionsError","message":"Tool call did not return a valid response (gram_err_001)","cause":"Expected instance of \`Response\` but got \`object\`"}"
`;

exports[`proxy good downstream fetch calls 1`] = `
"HTTP/1.1 200 OK
content-length: 28
content-type: application/json

{"name":"Finn","type":"cat"}"
`;

exports[`valid tool call adds custom headers 1`] = `
"HTTP/1.1 200 
content-type: application/json
x-next-cursor: 4

{"currentCursor":"1","items":[{"id":1,"name":"Product A","price":10},{"id":2,"name":"Product B","price":15.5},{"id":3,"name":"Product C","price":7.25}]}"
`;

exports[`valid tool call explicit 4XX 1`] = `
"HTTP/1.1 400 
content-type: application/json

{"error":"Invalid amount"}"
`;

exports[`valid tool call explicit 4XX 2`] = `
"HTTP/1.1 400 
content-type: application/json

{"error":"Invalid amount"}"
`;

exports[`valid tool call with args 1`] = `
"HTTP/1.1 200 
content-type: application/json

{"location":"San Francisco","temperature":"20C"}"
`;

exports[`valid tool call without args 1`] = `
"HTTP/1.1 200 
content-type: text/plain;charset=UTF-8

{"response":"pong"}"
`;
